packages/menu/src/Menu.ts

```diff
@@ -201,30 +370,90 @@ export class Menu extends SpectrumElement {
             item = this.menuItems[index] as MenuItem;
         }
         index = Math.max(index, 0);
+        const selectedItemsMap = new Map<MenuItem, boolean>();
+        const selected: string[] = [];
+        const selectedItems: MenuItem[] = [];
         this.menuItems.forEach((item, i) => {
+            if (item.selected) {
+                selectedItemsMap.set(item, true);
+                selected.push(item.value);
+                selectedItems.push(item);
+            }
             if (i !== index) {
                 item.focused = false;
             }
         });
+        this.selectedItemsMap = selectedItemsMap;
+        this.selected = selected;
+        this.selectedItems = selectedItems;
         this.focusedItemIndex = index;
         this.focusInItemIndex = index;
     }
 
-    private prepItems = (): void => {
-        this.menuItems = [
-            ...this.querySelectorAll(`[role="${this.childRole}"]`),
-        ] as MenuItem[];
-        if (!this.menuItems || this.menuItems.length === 0) {
-            return;
+    // debounce update this.menuItems so they're in DOM order
+    private async updateMenuItems(): Promise<void> {
+        this.itemsChanged = true;
+
+        return new Promise((resolve) => {
+            // debounce so we only actually update once
+            requestAnimationFrame(() => {
+                if (!this.itemsChanged) {
+                    resolve();
+                }
+
+                this.menuItems = [];
+
+                const slotElements =
+                    this.menuSlot.assignedElements({ flatten: true }) || [];
+                for (const slotElement of slotElements) {
+                    const childItems: MenuItem[] =
+                        slotElement instanceof MenuItem
+                            ? [slotElement as MenuItem]
+                            : ([
+                                  ...slotElement.querySelectorAll(
+                                      `[role="${this.childRole}"]`
+                                  ),
+                              ] as MenuItem[]);
+                    for (const childItem of childItems) {
+                        if (this.menuItemSet.has(childItem)) {
+                            this.menuItems.push(childItem);
+                        }
+                    }
+                }
+
+                this.updateSelectedItemIndex();
+                this.updateItemFocus();
+                this.itemsChanged = false;
+                resolve();
+            });
+        });
+    }
+
+    private addItem(item: MenuItem) {
+        this.menuItemSet.add(item);
+        this.updateMenuItems();
+    }
```

Comment:
This and the `updateMenuItems` method are the bulk of the new management logic. While there's a lot more to clean up in this PR, it'd be helpful to get some feedback on this approach.

Reply:
IIUC, the `this.menuItems` array that is built in `this.updateMenuItems()` only needs to be maintained in the original `sp-menu` descendent, right? Like in this case:
```
<sp-menu> // I'm a menu.
  <sp-menu-group> // I own a menu
    <sp-menu-item></sp-menu-item>
  </sp-menu-group>
  <sp-menu-group> // I also own a meni
    <sp-menu-item></sp-menu-item>
  </sp-menu-group>
  <sp-menu> // I'm also a menu
    <sp-menu-item></sp-menu-item>
  </sp-menu>
</sp-menu>
```
While _technically_ there would be 4 `sp-menu` elements in the various shadow trees, only the one at the root would need to coalesce this data in order to manage focus/arrow keys/etc. Do we need additional logic to skip that work, or am I missing something?

Reply:
If a sub-menu manages it's own selection things get trickier, since we still want arrow keys to behave sensibly. I haven't addressed that yet (realized this case right after I pushed up yesterday). So tweaking your example slightly:

```
<sp-menu> // I'm a menu.
  <sp-menu-group> // I own a menu
    <sp-menu-item></sp-menu-item>
  </sp-menu-group>
  <sp-menu-group selects="single"> // I'm a menu that manages my own selection
    <sp-menu-item></sp-menu-item>
  </sp-menu-group>
  <sp-menu> // I'm also a menu
    <sp-menu-item></sp-menu-item>
  </sp-menu>
</sp-menu>
```

Reply:
One question I have here is whether I need to manage 2 lists. One for focus (every item the menu is a parent of), and one for just the items the menu directly manages.

Reply:
This is actually what I was thinking. There are _ways_ to not do this, but I'm pretty certain that while the selection context is handled locally, the focus management has to happen globally. There is likely some tricky accessibility concerns with how the screen reader relates to the data here...

Reply:
You could also have a more complicated single list?
```
type ChildItem = {
    element: MenuItem, // reference to DOM element
    managed: boolean, // whether this item is part of the selection pool
    active: boolean, // whether this is the "focused" item or not
}

type ChildItems = ChildItem[];
```

Reply:
Okay, this is helpful!

Would only the top-level parent menu need to manage the focus list (at least until we support nested menus)?

Reply:
I believe so.

For all intents and purposes, each nested menu will act as its own root: https://webcomponents.dev/edit/Aze4atcpZc7EsxqnCBve/src/index.ts anyways.